src/main/java/javaslang/collection/HashSet.java

```diff
@@ -0,0 +1,515 @@
+
+    @SuppressWarnings("unchecked")
+    public static <T> HashSet<T> empty() {
+        return (HashSet<T>) INSTANCE;
+    }
+
+    /**
+     * Returns a {@link java.util.stream.Collector} which may be used in conjunction with
+     * {@link java.util.stream.Stream#collect(java.util.stream.Collector)} to obtain a {@link javaslang.collection.Set}s.
+     *
+     * @param <T> Component type of the List.
+     * @return A javaslang.collection.List Collector.
+     */
+    public static <T> Collector<T, ArrayList<T>, Set<T>> collector() {
+        final Supplier<ArrayList<T>> supplier = ArrayList::new;
+        final BiConsumer<ArrayList<T>, T> accumulator = ArrayList::add;
+        final BinaryOperator<ArrayList<T>> combiner = (left, right) -> {
+            left.addAll(right);
+            return left;
+        };
+        final Function<ArrayList<T>, Set<T>> finisher = HashSet::ofAll;
+        return Collector.of(supplier, accumulator, combiner, finisher);
+    }
+
+    /**
+     * Returns a singleton {@code HashSet}, i.e. a {@code HashSet} of one element.
+     *
+     * @param element An element.
+     * @param <T>     The component type
+     * @return A new HashSet instance containing the given element
+     */
+    static <T> HashSet<T> of(T element) {
+        return HashSet.<T>empty().add(element);
+    }
+
+    /**
+     * <p>
+     * Creates a HashSet of the given elements.
+     * </p>
+     *
+     * <pre>
+     * <code>  HashSet.of(1, 2, 3, 4)
+     * </pre>
+     *
+     * @param <T>      Component type of the HashSet.
+     * @param elements Zero or more elements.
+     * @return A set containing the given elements.
+     * @throws NullPointerException if {@code elements} is null
+     */
+    @SafeVarargs
+    static <T> HashSet<T> of(T... elements) {
+        Objects.requireNonNull(elements, "elements is null");
+        HashSet<T> result = HashSet.empty();
+        for (int i = 0; i < elements.length; i++) {
+            result = result.add(elements[i]);
+        }
+        return result;
+    }
+
+    /**
+     * Creates a HashSet of the given entries.
+     *
+     * @param entries Set entries
+     * @param <T>     The value type
+     * @return A new HashSet containing the given entries
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> HashSet<T> ofAll(Iterable<? extends T> entries) {
+        Objects.requireNonNull(entries, "entries is null");
+        if (entries instanceof HashSet) {
+            return (HashSet<T>) entries;
+        } else {
+            HashArrayMappedTrie<T, Object> tree = HashArrayMappedTrie.empty();
+            for (T entry : entries) {
+                tree = tree.put(entry, O);
+            }
+            return tree.isEmpty() ? empty() : new HashSet<>(tree);
+        }
+    }
+
+    private final static Object O = new Object();
+
+    private final HashArrayMappedTrie<T, Object> tree;
+    private final Lazy<List<T>> list;
+    private final Lazy<Integer> hash;
+
```

Comment:
With `list` we trade space against performance here. This is nice regarding performance. I wonder how Scala does this for Map/Set... I'm just curious.